common/datasets/sms_wsj/returnn_datasets.py

```diff
@@ -0,0 +1,617 @@
+        """
+        Returns data for sequence index.
+        """
+        if self._use_buffer:
+            assert (
+                seq_idx in self._buffer
+            ), f"seq_idx {seq_idx} not in buffer. Available keys are {self._buffer.keys()}"
+            return self._buffer[seq_idx]
+        else:
+            return self._ds[seq_idx]
+
+    def get_seq_tag(self, seq_idx: int) -> str:
+        """
+        Returns tag for the sequence of the given index, default is 'seq-{seq_idx}'.
+        """
+        return str(self._get_seq_by_idx(seq_idx).get("seq_tag", f"seq-{seq_idx}"))
+
+    def get_seq_len(self, seq_idx: int) -> int:
+        """
+        Returns length of the sequence of the given index
+        """
+        try:
+            return int(self._get_seq_by_idx(seq_idx)["seq_len"])
+        except KeyError:
+            raise OptionalNotImplementedError
+
+    def get_seq_length_for_keys(self, seq_idx: int) -> NumbersDict:
+        """
+        Returns sequence length for all data/target keys.
+        """
+        data = self[seq_idx]
+        d = {k: v.size for k, v in data.items()}
+        for update_key in ["data", "target_signals"]:
+            if update_key in d and "seq_len" in data:
+                d[update_key] = int(data["seq_len"])
+        return NumbersDict(d)
+
+    def update_buffer(self, seq_idx: int, pop_seqs: bool = True):
+        """
+        :param int seq_idx:
+        :param bool pop_seqs: if True, pop sequences from buffer that are outside buffer range
+        """
+        if not self._use_buffer:
+            return
+
+        # debugging information
+        keys = list(self._buffer.keys()) or [0]
+        if not (min(keys) <= seq_idx <= max(keys)):
+            print(
+                f"WARNING: seq_idx {seq_idx} outside range of keys: {self._buffer.keys()}",
+                file=returnn_log.v5,
+            )
+
+        # add sequences
+        for idx in range(seq_idx, min(seq_idx + self._buffer.max_size // 2, len(self))):
+            if idx not in self._buffer:
+                self._buffer[idx] = next(self._ds_iterator)
+            if idx == len(self) - 1 and 0 not in self._buffer:
+                print(f"Reached end of dataset, reset iterator", file=returnn_log.v4)
+                try:
+                    next(self._ds_iterator)
+                except StopIteration:
+                    pass
+                else:
+                    print(
+                        "WARNING: reached final index of dataset, but iterator has more sequences. "
+                        "Maybe the training was restarted from an epoch > 1?",
+                        file=returnn_log.v3,
+                    )
+                print(
+                    f"Current buffer indices: {self._buffer.keys()}",
+                    file=returnn_log.v5,
+                )
+                self._ds_iterator = iter(self._ds)
+                for idx_ in range(min(self._buffer.max_size // 2, len(self))):
```

Comment:
I like the idea of having a class for the buffer, but I have faced a problem with the `update_buffer` function, more precisely, the automatic popping of the oldest inserted element. 
When the end of the dataset is reached, the buffer is "refilled" with `self._buffer.max_size // 2` sequences from the beginning of the dataset and the oldest `self._buffer.max_size // 2` sequences are popped. However, when working with batches, this may cause that sequences which are needed for the batch are removed from the buffer too soon.

This could maybe be avoided by filling one sequence from the beginning of the dataset at a time. Maybe something like:
 `for idx_ in range((seq_idx + self._buffer.max_size // 2) - len(self) + 1)`

Reply:
The issue with that might be that you only get here if `idx == len(self) - 1 and 0 not in self._buffer`. So probably when `seq_idx + self._buffer.max_size // 2 == len(self) - 1`. So you would have `range(0)`, right? A simpler thing might be to just increase the buffer size in your case.

Reply:
I don't think increasing the buffer size will help because the oldest `self._buffer.max_size // 2` sequences are popped, no matter if the buffer size is 40 or 400, you will always loose the first half of sequences in the buffer with that loop.

So tell me if I'm wrong, but due to `min(seq_idx + self._buffer.max_size // 2, len(self))`, we get there when `seq_idx + self._buffer.max_size // 2 -1 == len(self) -1` and not `seq_idx + self._buffer.max_size // 2 == len(self) - 1`. That's why I added `+1`, so we have at least `range(1)`. However you are right that the condition `0 not in self._buffer` causes a problem

Reply:
Sure, the oldest `self._buffer.max_size // 2` sequences are always popped, but if the buffer size is large enough, the remaining old sequences are sufficient to cover everything that you need for your remaining batches, no?

Reply:
I think it's easier to give an example:
Let's say `len(self)` is 125 and my buffer size is 40. My batch begins at sequence 100 and will contain 10 sequences. Now, at some point, `update_buffer` is called with `seq_idx=105`. 
Then, in the first part, the buffer is filled with `85,...,105,...,124` (because `124 == seq_idx + self._buffer.max_size // 2 -1`, so we do `for idx in range(105, 125)`). Now, the condition `if idx == len(self) - 1 and 0 not in self._buffer` is fulfilled (because `124 == len(self) - 1`). `for idx_ in range(min(self._buffer.max_size // 2, len(self)))` fills the buffer with indices 0 to 19. As a result, the first 20 elements of the buffer are popped, so the buffer is now filled with elements `105,...,124,0,...,19`. However, for the batch we need sequences 100 to 104, which are not in the buffer anymore.
That means, everything before sequence 105 is removed, there are no old sequences remaining. With an increased batch size, the problem would just appear earlier.
Again, tell me if I am wrong, maybe I am doing a mistake here.

Reply:
Yes I agree. In general as soon as update_buffer is called on `seq_idx = len(self) - buffer_size // 2` the start of the dataset is added and the buffer is now filled with sequences `[len(self) - buffer_size // 2 ... len(self) - 1] + [0 ... buffer_size // 2 - 1]` so this only keeps the current minibatch intact if `seq_idx` happens to be the first sequence in the minibatch, regardless of buffer_size. In the example of @larissakl it only runs successfully if the batch begins at sequence 105.

I ran into this error as well and changed the instances of `// 2` to `// 3` locally which seems to give enough room for old sequences but maybe the safer solution is to only add `seq_idx + buffer_size // 2 - len(self)` sequences from the dataset start. For example when `update_buffer(118)` is called, the first `13` sequences would be added such that the buffer contains `98,...,118,...,124,0,...,12` afterwards.

BTW I think we should add some documentation that the buffer size should be set large enough so that the number of sequences in a mini-batch never exceeds `buffer_size // 2`; otherwise the training will crash.

Reply:
> I ran into this error as well and changed the instances of `// 2` to `// 3` locally which seems to give enough room for old sequences but maybe the safer solution is to only add `seq_idx + buffer_size // 2 - len(self)` sequences from the dataset start. For example when `update_buffer(118)` is called, the first `13` sequences would be added such that the buffer contains `98,...,118,...,124,0,...,12` afterwards.

That's what I meant by `for idx_ in range((seq_idx + self._buffer.max_size // 2) - len(self) + 1)`. I tried the following and it seems to work:

```
        for idx in range(seq_idx, min(seq_idx + self._buffer.max_size // 2, len(self))):
            if idx not in self._buffer:
                self._buffer[idx] = next(self._ds_iterator)
            if idx == len(self) - 1:
                if 0 not in self._buffer:
                    print(f"Reached end of dataset, reset iterator", file=returnn_log.v4)
                    try:
                        next(self._ds_iterator)
                    except StopIteration:
                        pass
                    else:
                        print(
                            "WARNING: reached final index of dataset, but iterator has more sequences. "
                            "Maybe the training was restarted from an epoch > 1?",
                            file=returnn_log.v3,
                        )
                    print(
                        f"Current buffer indices: {self._buffer.keys()}",
                        file=returnn_log.v5,
                    )
                    self._ds_iterator2 = iter(self._ds)
                for idx_ in range((seq_idx + self._buffer.max_size // 2) - len(self) + 1):
                    if idx_ not in self._buffer:
                        self._buffer[idx_] = next(self._ds_iterator2)
                    if seq_idx == len(self) - 1:
                        self._ds_iterator = self._ds_iterator2
```

It realizes the idea of only adding `seq_idx + buffer_size // 2 - len(self)` sequences from the dataset start. The iterator is only reset if 0 is not in the buffer. If this happens, we have to work with two iterators in parallel, one at the end of the dataset (until the last sequence is reached) and one at the start.

Reply:
I pushed a fix which also simplifies this a bit. When we reach the last index (`len(self) - 1`) for the first time, we reset the iterator and add the index 0). Then, we always add one index more at the beginning of the dataset by using `idx % len(self)` instead of just `idx` capped at `len(self)`.

Reply:
Please go ahead and test that, for my local test case the issue is resolved.

Reply:
> BTW I think we should add some documentation that the buffer size should be set large enough so that the number of sequences in a mini-batch never exceeds buffer_size // 2; otherwise the training will crash.

I added that in the docstring, you can have a look whether you find that sufficient.